File: kurrent_skills/templates/nodejs/projection.js
/**
 * KurrentDB Node.js In-Memory Projection Example
 * Demonstrates: Subscribe to events, build state, track checkpoint
 */

import {
    KurrentDBClient,
    jsonEvent,
    START,
    excludeSystemEvents,
} from "@kurrent/kurrentdb-client";
import { v4 as uuid } from "uuid";

// === MINIMAL PROJECTION FRAMEWORK ===
class Projection {
    constructor(name) {
        this.name = name;
        this.state = new Map();
        this.checkpoint = null;
        this.handlers = new Map();
    }

    on(eventType, handler) {
        this.handlers.set(eventType, handler);
        return this;
    }

    get(id) {
        return this.state.get(id);
    }

    getAll() {
        return Object.fromEntries(this.state);
    }

    apply(resolvedEvent) {
        const event = resolvedEvent.event;
        if (!event) return false;

        const handler = this.handlers.get(event.type);
        if (handler) {
            const streamId = event.streamId;
            const current = this.state.get(streamId) || {};
            const newState = handler(current, event.data, event);
            this.state.set(streamId, newState);
            this.checkpoint = resolvedEvent.commitPosition;
            return true;
        }
        return false;
    }
}

// === SETUP ===
const connectionString = process.env.KURRENTDB_CONNECTION_STRING || "kurrentdb://localhost:2113?tls=false";
const client = KurrentDBClient.connectionString(connectionString);
console.log(`Connected to KurrentDB at ${connectionString}`);

// === DEFINE PROJECTION ===
const orderProjection = new Projection("OrderSummary")
    .on("OrderCreated", (state, data) => ({
        orderId: data.orderId,
        customerId: data.customerId,
        amount: data.amount,
        status: "created",
        items: []
    }))
    .on("ItemAdded", (state, data) => ({
        ...state,
        items: [...(state.items || []), data.item],
        amount: state.amount + data.price
    }))
    .on("OrderShipped", (state, data) => ({
        ...state,
        status: "shipped",
        shippedAt: data.shippedAt
    }))
    .on("OrderCompleted", (state) => ({
        ...state,
        status: "completed"
    }));

// === TEST: Append test events ===
console.log("\n=== Appending test events ===");

const orderId1 = uuid();
const orderId2 = uuid();
const stream1 = `order-${orderId1}`;
const stream2 = `order-${orderId2}`;

// Order 1: Created -> ItemAdded -> Shipped -> Completed
await client.appendToStream(stream1, [
    jsonEvent({ type: "OrderCreated", data: { orderId: orderId1, customerId: "cust-1", amount: 100 } }),
]);
await client.appendToStream(stream1, [
    jsonEvent({ type: "ItemAdded", data: { item: "Widget", price: 25 } }),
]);
await client.appendToStream(stream1, [
    jsonEvent({ type: "OrderShipped", data: { shippedAt: new Date().toISOString() } }),
]);
await client.appendToStream(stream1, [
    jsonEvent({ type: "OrderCompleted", data: {} }),
]);

// Order 2: Created -> ItemAdded (still pending)
await client.appendToStream(stream2, [
    jsonEvent({ type: "OrderCreated", data: { orderId: orderId2, customerId: "cust-2", amount: 50 } }),
]);
await client.appendToStream(stream2, [
    jsonEvent({ type: "ItemAdded", data: { item: "Gadget", price: 30 } }),
]);

console.log(`Created order streams: ${stream1}, ${stream2}`);

// === RUN PROJECTION ===
console.log("\n=== Running projection ===");

const subscription = client.subscribeToAll({
    fromPosition: START,
    filter: excludeSystemEvents(),
});

const targetStreams = new Set([stream1, stream2]);
const targetEventsCount = { [stream1]: 0, [stream2]: 0 };
const expectedCounts = { [stream1]: 4, [stream2]: 2 }; // Order 1: 4 events, Order 2: 2 events

for await (const event of subscription) {
    const streamId = event.event?.streamId;

    if (orderProjection.apply(event)) {
        console.log(`  Processed: ${event.event.type} on ${streamId}`);

        if (targetStreams.has(streamId)) {
            targetEventsCount[streamId]++;
        }
    }

    // Stop when we've processed all our test events
    if (targetEventsCount[stream1] >= expectedCounts[stream1] &&
        targetEventsCount[stream2] >= expectedCounts[stream2]) {
        break;
    }

    // Safety limit
    const total = targetEventsCount[stream1] + targetEventsCount[stream2];
    if (total > 200) break;
}

// === VERIFY RESULTS ===
console.log("\n=== Projection Results ===");

const order1State = orderProjection.get(stream1);
const order2State = orderProjection.get(stream2);

console.log(`\nOrder 1 (${stream1}):`);
console.log(JSON.stringify(order1State, null, 2));

console.log(`\nOrder 2 (${stream2}):`);
console.log(JSON.stringify(order2State, null, 2));

// === ASSERTIONS ===
console.log("\n=== Running assertions ===");

let passed = true;

// Order 1 assertions
if (order1State?.status !== "completed") {
    console.error(`FAIL: Order 1 status should be 'completed', got '${order1State?.status}'`);
    passed = false;
}
if (order1State?.amount !== 125) {
    console.error(`FAIL: Order 1 amount should be 125, got ${order1State?.amount}`);
    passed = false;
}
if (order1State?.items?.length !== 1) {
    console.error(`FAIL: Order 1 should have 1 item, got ${order1State?.items?.length}`);
    passed = false;
}

// Order 2 assertions
if (order2State?.status !== "created") {
    console.error(`FAIL: Order 2 status should be 'created', got '${order2State?.status}'`);
    passed = false;
}
if (order2State?.amount !== 80) {
    console.error(`FAIL: Order 2 amount should be 80, got ${order2State?.amount}`);
    passed = false;
}

// Checkpoint assertion
if (!orderProjection.checkpoint) {
    console.error("FAIL: Checkpoint should be set");
    passed = false;
}

if (passed) {
    console.log("\nAll projection tests passed!");
} else {
    console.error("\nSome tests failed!");
    process.exit(1);
}

process.exit(0);

File: kurrent_skills/templates/rust/src/projection.rs
//! KurrentDB Rust In-Memory Projection Example
//! Demonstrates: Subscribe to events, build state, track checkpoint

use kurrentdb::{
    Client, EventData, Position, SubscribeToAllOptions, SubscriptionFilter,
};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;
use std::env;
use uuid::Uuid;

// === MINIMAL PROJECTION FRAMEWORK ===

type EventHandler = Box<dyn Fn(&Value, &Value) -> Value + Send + Sync>;

pub struct Projection {
    pub name: String,
    pub state: HashMap<String, Value>,
    pub checkpoint: Option<Position>,
    handlers: HashMap<String, EventHandler>,
}

impl Projection {
    pub fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
            state: HashMap::new(),
            checkpoint: None,
            handlers: HashMap::new(),
        }
    }

    pub fn on<F>(mut self, event_type: &str, handler: F) -> Self
    where
        F: Fn(&Value, &Value) -> Value + Send + Sync + 'static,
    {
        self.handlers.insert(event_type.to_string(), Box::new(handler));
        self
    }

    pub fn get(&self, stream_id: &str) -> Option<&Value> {
        self.state.get(stream_id)
    }

    pub fn apply(&mut self, stream_id: &str, event_type: &str, data: &[u8], position: Position) -> bool {
        if let Some(handler) = self.handlers.get(event_type) {
            let current = self.state.get(stream_id).cloned().unwrap_or(json!({}));
            let event_data: Value = serde_json::from_slice(data).unwrap_or(json!({}));
            let new_state = handler(&current, &event_data);
            self.state.insert(stream_id.to_string(), new_state);
            self.checkpoint = Some(position);
            return true;
        }
        false
    }
}

// === ORDER EVENTS ===

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct OrderCreated {
    order_id: String,
    customer_id: String,
    amount: f64,
}

#[derive(Debug, Serialize, Deserialize)]
struct ItemAdded {
    item: String,
    price: f64,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct OrderShipped {
    shipped_at: String,
}

pub async fn run_projection() -> Result<(), Box<dyn std::error::Error>> {
    // === SETUP ===
    let connection_string = env::var("KURRENTDB_CONNECTION_STRING")
        .unwrap_or_else(|_| "kurrentdb://localhost:2113?tls=false".to_string());

    let settings = connection_string.parse()?;
    let client = Client::new(settings)?;

    println!("Connected to KurrentDB at {}", connection_string);

    // === DEFINE PROJECTION ===
    let mut order_projection = Projection::new("OrderSummary")
        .on("OrderCreated", |_state, data| {
            json!({
                "orderId": data.get("orderId"),
                "customerId": data.get("customerId"),
                "amount": data.get("amount"),
                "status": "created",
                "items": []
            })
        })
        .on("ItemAdded", |state, data| {
            let mut new_state = state.clone();
            if let Some(items) = new_state.get_mut("items") {
                if let Some(arr) = items.as_array_mut() {
                    arr.push(json!(data.get("item")));
                }
            }
            if let (Some(current), Some(price)) = (
                new_state.get("amount").and_then(|v| v.as_f64()),
                data.get("price").and_then(|v| v.as_f64()),
            ) {
                new_state["amount"] = json!(current + price);
            }
            new_state
        })
        .on("OrderShipped", |state, data| {
            let mut new_state = state.clone();
            new_state["status"] = json!("shipped");
            new_state["shippedAt"] = json!(data.get("shippedAt"));
            new_state
        })
        .on("OrderCompleted", |state, _data| {
            let mut new_state = state.clone();
            new_state["status"] = json!("completed");
            new_state
        });

    // === TEST: Append test events ===
    println!("\n=== Appending test events ===");

    let order_id_1 = Uuid::new_v4().to_string();
    let order_id_2 = Uuid::new_v4().to_string();
    let stream_1 = format!("order-{}", order_id_1);
    let stream_2 = format!("order-{}", order_id_2);

    // Order 1: Created -> ItemAdded -> Shipped -> Completed
    let event = EventData::json("OrderCreated", &OrderCreated {
        order_id: order_id_1.clone(),
        customer_id: "cust-1".to_string(),
        amount: 100.0,
    })?.id(Uuid::new_v4());
    client.append_to_stream(stream_1.clone(), &Default::default(), event).await?;

    let event = EventData::json("ItemAdded", &ItemAdded {
        item: "Widget".to_string(),
        price: 25.0,
    })?.id(Uuid::new_v4());
    client.append_to_stream(stream_1.clone(), &Default::default(), event).await?;

    let event = EventData::json("OrderShipped", &OrderShipped {
        shipped_at: "2024-01-15T10:00:00Z".to_string(),
    })?.id(Uuid::new_v4());
    client.append_to_stream(stream_1.clone(), &Default::default(), event).await?;

    let event = EventData::json("OrderCompleted", &json!({}))?.id(Uuid::new_v4());
    client.append_to_stream(stream_1.clone(), &Default::default(), event).await?;

    // Order 2: Created -> ItemAdded (still pending)
    let event = EventData::json("OrderCreated", &OrderCreated {
        order_id: order_id_2.clone(),
        customer_id: "cust-2".to_string(),
        amount: 50.0,
    })?.id(Uuid::new_v4());
    client.append_to_stream(stream_2.clone(), &Default::default(), event).await?;

    let event = EventData::json("ItemAdded", &ItemAdded {
        item: "Gadget".to_string(),
        price: 30.0,
    })?.id(Uuid::new_v4());
    client.append_to_stream(stream_2.clone(), &Default::default(), event).await?;

    println!("Created order streams: {}, {}", stream_1, stream_2);

    // === RUN PROJECTION ===
    println!("\n=== Running projection ===");

    let filter = SubscriptionFilter::on_event_type().exclude_system_events();
    let options = SubscribeToAllOptions::default().filter(filter);
    let mut subscription = client.subscribe_to_all(&options).await;

    let mut processed_count = 0;
    let mut target_events_count: HashMap<String, i32> = HashMap::new();
    target_events_count.insert(stream_1.clone(), 0);
    target_events_count.insert(stream_2.clone(), 0);
    let expected_counts: HashMap<String, i32> =
        [(stream_1.clone(), 4), (stream_2.clone(), 2)].into_iter().collect();

    loop {
        match subscription.next().await {
            Ok(event) => {
                let original = event.get_original_event();
                let stream_id = original.stream_id();
                let event_type = &original.event_type;
                let position = original.position;

                if order_projection.apply(stream_id, event_type, &original.data, position) {
                    processed_count += 1;
                    println!("  Processed: {} on {}", event_type, stream_id);

                    if let Some(count) = target_events_count.get_mut(stream_id) {
                        *count += 1;
                    }
                }

                // Stop when we've processed all test events for both streams
                let stream1_done = target_events_count.get(&stream_1).unwrap_or(&0) >= expected_counts.get(&stream_1).unwrap_or(&4);
                let stream2_done = target_events_count.get(&stream_2).unwrap_or(&0) >= expected_counts.get(&stream_2).unwrap_or(&2);
                if stream1_done && stream2_done {
                    break;
                }

                // Safety limit
                if processed_count > 200 {
                    break;
                }
            }
            Err(e) => {
                println!("Subscription error: {}", e);
                break;
            }
        }
    }

    // === VERIFY RESULTS ===
    println!("\n=== Projection Results ===");

    let order1_state = order_projection.get(&stream_1);
    let order2_state = order_projection.get(&stream_2);

    println!("\nOrder 1 ({}):", stream_1);
    println!("{}", serde_json::to_string_pretty(&order1_state)?);

    println!("\nOrder 2 ({}):", stream_2);
    println!("{}", serde_json::to_string_pretty(&order2_state)?);

    // === ASSERTIONS ===
    println!("\n=== Running assertions ===");

    let mut passed = true;

    // Order 1 assertions
    if let Some(state) = order1_state {
        if state.get("status").and_then(|v| v.as_str()) != Some("completed") {
            println!("FAIL: Order 1 status should be 'completed', got {:?}", state.get("status"));
            passed = false;
        }
        if state.get("amount").and_then(|v| v.as_f64()) != Some(125.0) {
            println!("FAIL: Order 1 amount should be 125, got {:?}", state.get("amount"));
            passed = false;
        }
        if state.get("items").and_then(|v| v.as_array()).map(|a| a.len()) != Some(1) {
            println!("FAIL: Order 1 should have 1 item");
            passed = false;
        }
    } else {
        println!("FAIL: Order 1 state not found");
        passed = false;
    }

    // Order 2 assertions
    if let Some(state) = order2_state {
        if state.get("status").and_then(|v| v.as_str()) != Some("created") {
            println!("FAIL: Order 2 status should be 'created', got {:?}", state.get("status"));
            passed = false;
        }
        if state.get("amount").and_then(|v| v.as_f64()) != Some(80.0) {
            println!("FAIL: Order 2 amount should be 80, got {:?}", state.get("amount"));
            passed = false;
        }
    } else {
        println!("FAIL: Order 2 state not found");
        passed = false;
    }

    // Checkpoint assertion
    if order_projection.checkpoint.is_none() {
        println!("FAIL: Checkpoint should be set");
        passed = false;
    }

    if passed {
        println!("\nAll projection tests passed!");
    } else {
        println!("\nSome tests failed!");
        std::process::exit(1);
    }

    Ok(())
}

// Uncomment to run as standalone:
// #[tokio::main]
// async fn main() -> Result<(), Box<dyn std::error::Error>> {
//     run_projection().await
// }
